### PR TITLE
dotnet: added shell completion scripts

### PR DESCRIPTION
**Pull request by gbtb**
## Description of changes

I've added shell completions scripts from official microsoft page to dotnet package

## Things done

I've checked bash and fish shells, completions works fine in there. When I tried to start zsh from `nix-shell -p zsh` it started complaining about missing config and stuff like that and completions didn't work. Would be nice if someone using zsh daily would test this PR. Other options would be to ship it as-is (my choice) or delete zsh completion from PR (if other reviewers decide so).

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by raphaelr — APPROVED** [review]
I tested the completions with `bash` and it works.

**Comment by gbtb** [comment]
Hey @NixOS/dotnet :wave: , could you please take a look?

**Comment by GGG-KILLER** [comment]
Hello, I'll check it. I'm using zsh so I can test with that as well

File: pkgs/development/compilers/dotnet/build-dotnet.nix [hunk 1]
```diff
@@ -82,6 +84,11 @@ mkCommon type rec {
     mv $out/ThirdPartyNotices.txt $out/share/doc/$pname/$version/
 
     ln -s $out/dotnet $out/bin/dotnet
+    # completions snippets taken from https://learn.microsoft.com/en-us/dotnet/core/tools/enable-tab-autocomplete
```
Comment (corngood) [inline pkgs/development/compilers/dotnet/build-dotnet.nix:87]: could we do this in `pkgs/development/compilers/dotnet/common.nix` to support the source-built things as well?

**Review by corngood — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/dotnet/build-dotnet.nix (hunk 1 above)
Comment (gbtb) [reply pkgs/development/compilers/dotnet/build-dotnet.nix:87]: :+1:  I think we can

**Comment by GGG-KILLER** [comment]
Can confirm it works on zsh:
image

File: pkgs/development/compilers/dotnet/build-dotnet.nix (hunk 1 above)
Comment (gbtb) [reply pkgs/development/compilers/dotnet/build-dotnet.nix]: I moved installShellFiles call to common.nix and tested that completions still works for dotnet_sdk package, as well as dotnetCorePackages.dotnet_8.sdk (which looks like the source-built package you've mentioned) .

**Review by corngood — APPROVED** [review]

**Review by GGG-KILLER — APPROVED** [review]

**Comment by corngood** [comment]
@gbtb I'm sorry, I tried and failed to resolve this conflict with the web editor.  would you mind rebasing it?

**Comment by gbtb** [comment]
> @gbtb I'm sorry, I tried and failed to resolve this conflict with the web editor. would you mind rebasing it?

I've rebased this PR on top of the latest master.

**Pull request merged**